File: java/Hot100/Substring/SubarraySumEqualsK.java
package Hot100.Substring;

import java.util.HashMap;
import java.util.Map;

/**
 * Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
 *
 * A subarray is a contiguous non-empty sequence of elements within an array.
 */
public class SubarraySumEqualsK {
    public int subarraySum(int[] nums, int k) {
        //current sum
        int preSum = 0;
        //record times of preSum
        Map<Integer, Integer> freq = new HashMap<>();
        freq.put(0, 1);
        int cnt = 0;

        for (int num : nums) {
            preSum += num;
            cnt += freq.getOrDefault(preSum - k, 0);
            freq.put(preSum, freq.getOrDefault(preSum, 0) + 1);
        }

        return cnt;
    }
}

File: python/substring/subarray_sum_equals_k.py
# Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
# A subarray is a contiguous non-empty sequence of elements within an array.

class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        prefix_sum = 0
        freq_map = {0: 1}  # Initialize with prefix sum 0 having one occurrence
        count = 0

        for num in nums:
            prefix_sum += num
            # Check if there is a prefix sum that when subtracted from the current prefix sum equals k
            count += freq_map.get(prefix_sum - k, 0)
            # Update the frequency map with the current prefix sum
            freq_map[prefix_sum] = freq_map.get(prefix_sum, 0) + 1

        return count
       